include/deal.II/distributed/shared_tria.h

```diff
@@ -106,6 +113,39 @@ namespace parallel
                                          const std::vector< CellData< dim > > &cells,
                                          const SubCellData &subcelldata);
 
+      /**
+       * Return auxilary vector subdomain_id_with_artificial .
```

Comment:
OK, seeing how you use this, how about this:
* Rename this function `get_true_subdomain_ids_of_cells()` and document it
* Remove the other function + corresponding variable
* In the place where you use it, first store the current subdomain ids into a local vector, then reset them using what you get from the `get_true_subdomain_ids_of_cells()` function
* Do what you need to do
* Restore subdomain ids from the local vector you created.